Integrate Image Streams

maneta opened the pull request:
ToDo: 

- [x] Check postCommit in the Build (https://github.com/3scale/apicast/pull/618)
- [x] set resource limits/requests
- [x] Update deployment documentation and place a `Makefile`
- [x] Prepare service for prometheus metrics.

Commit:
Introducing Imagestreams

Plus separating templates for easier Deployment

mikz commented on openshift/01-apicast-cloud-hosted-imagestream.yml:
```diff
@@ -0,0 +1,14 @@
+  tags:
+  - name: latest
+    annotations:
+      openshift.io/display-name: Apicast Cloud Hosted (latest) 
```
`s/Apicast/APIcast/` 😉

maneta replied:
Oops I did it again :p

mikz commented on openshift/02-apicast-builder-imagestream.yml:
```diff
@@ -0,0 +1,12 @@
+    app: apicast
+spec:
+  tags:
+  - name: builder
```
Maybe add the branch name to the tag name? So we know it is "master" and not some tag?

maneta replied:
Good Idea 👍

mikz commented on openshift/04-deployment-template.yml:
```diff
@@ -77,10 +77,12 @@ objects:
+          image: apicast-cloud-hosted:${RELEASE_REF}
           imagePullPolicy: IfNotPresent
-          name: apicast
+          name: apicast-${RELEASE_REF}
```
There should be no need for containers to have a suffix no?

maneta replied:
yes sure

mikz commented on openshift/04-deployment-template.yml:
```diff
@@ -77,10 +77,12 @@ objects:
-          image: "${APICAST_IMAGE}"
+            value: "http://apicast-mapping-service-${RELEASE_REF}/config"
+          - name: APICAST_OIDC_LOG_LEVEL
+            value: "${APICAST_OIDC_LOG_LEVEL}"
```
I think this does not need to be a Parameter.

maneta replied:
Ok So I will place the env var with notice as default.

mikz commented on openshift/02-apicast-builder-imagestream.yml:
```diff
@@ -0,0 +1,12 @@
+  - name: builder
+    from:
+      kind: DockerImage
+      name: quay.io/3scale/apicast:master-builder
```
Shuld this use https://docs.openshift.com/container-platform/3.6/architecture/core_concepts/builds_and_image_streams.html#image-stream-mappings-working-periodic ?

mikz replied:
That could be set as https://docs.openshift.com/container-platform/3.6/rest_api/openshift_v1.html#v1-tagimportpolicy

maneta replied:
Yes I think it is a good idea.

Commit:
PR #4 review requests.

Commit:
Introducing Resource Limits

Same as APICast

maneta commented on openshift/04-deployment-template.yml:
```diff
@@ -103,6 +110,13 @@ objects:
           - containerPort: 8090
             name: management
             protocol: TCP
+          resources:
```
@mikz Should we set the `APICAST_WORKERS` too?

mikz replied:
Nope. APIcast autodetects number of requested CPU cores: https://github.com/3scale/apicast/pull/600

maneta replied:
ok

mikz commented on openshift/02-apicast-builder-imagestream.yml:
```diff
@@ -0,0 +1,12 @@
+  - name: master-builder
+    from:
+      kind: DockerImage
+      name: quay.io/3scale/apicast:master-builder
```
Would be good to have a comment pointing to how is this refreshed.
https://github.com/3scale/apicast-cloud-hosted/pull/4#discussion_r169143362

maneta replied:
TBH, I don't know how `quay.io/3scale/apicast:master-builder` is refreshed either...  Or are you talking about the scheduled imports from the registry?

mikz replied:
Yep. Scheduled imports. 

Those quay images are built when master branch is updated.

maneta replied:
Done in 287e33c18f82a14d84e86f17a83dc1c5239aabcd

mikz commented on openshift/04-deployment-template.yml:
```diff
@@ -101,29 +110,43 @@ objects:
+        imageChangeParams:
+          automatic: true
+          containerNames:
+            - apicast-${RELEASE_REF} 
```
This looks wrong.

maneta replied:
yep.

Commit:
Scheduled Import builder image

The ImageStream will immport the builder imager periodically. This configuration is clusterwide but the default is 15 minutes.

Commit:
Makefile for deployment and Documentation Update

mikz commented on openshift/Makefile:
```diff
@@ -0,0 +1,28 @@
+	oc new-app -f $(THISDIR_PATH)/04-deployment-template.yml \
+		-p RELEASE_REF=${RELEASE_REF} \
+		-p ENVIRONMENT=${ENVIRONMENT} \
+		-p CACHE_TTL=${CACHE_TTL} 
```
Will this crash when `CACHE_TTL` is empty? It should.

maneta replied:
yes it crashes because there is no predefined value in the template.

Actually it brakes without `RELEASE_REF` AND `ENVIRONMENT` also.

The actuall error message:

```bash
oc new-app -f /Users/dcesario/src/apicast-cloud-hosted/openshift/04-deployment-template.yml \
		-p RELEASE_REF=test-dani3 \
		-p ENVIRONMENT=staging \
		-p CACHE_TTL=
error: error processing template "apicast-staging/apicast-cloud-hosted-deployment": Template "apicast-cloud-hosted-deployment" is invalid: template.parameters[2]: Required value: template.parameters[2]: parameter CACHE_TTL is required and must be specified
make: *** [deploy] Error 1

mikz replied:
👍

maneta commented on openshift/03-build-config.yml:
```diff
@@ -7,18 +7,16 @@ metadata:
-      - '--daemon'
-    command:
-      - bin/apicast
+  #postCommit:
```
@mikz Should we remove the `postCommit` ?  The Deployment is working without it, but I'm not sure how to solve it...

mikz replied:
`postCommit` is good. It verifies the image will boot. Otherwise you have no idea until you try to deploy it (and fail).

maneta replied:
Ok you said something about Apicasy and the `RESOLVER` env Variable I will take a look at it

maneta replied:
@mikz So the problem with the postCommit is that the openshift builder is setting the `resolv.conf` file like that:

```
# nameserver updated by /etc/NetworkManager/dispatcher.d/99-origin-dns.sh
# Generated by NetworkManager
search ec2.internal cluster.local
nameserver 10.0.101.97
```
The problem is the first line, looks like the `parse_nameservers` function in APIcast https://github.com/3scale/apicast/blob/5c2dc6bbf9295b5e7891fb7bb52e32abcf763972/gateway/src/resty/resolver.lua#L105 is matching with the commented `nameserver` in the file.

The actual error thrown by nginx is:

`nginx: [emerg] invalid IPv6 address in resolver "[domain]:53" in /tmp/lua_c5D5o5:45`

I have reproduced it locally.

mikz replied:
Yep. Looks like: https://github.com/3scale/apicast/blob/5c2dc6bbf9295b5e7891fb7bb52e32abcf763972/gateway/src/resty/resolver.lua#L135-L140


This should match only from start of the line and ignore comments. We probably need better parsing anyway. 

Good catch!

maneta replied:
Corrected in https://github.com/3scale/apicast/pull/618

maneta commented:
@mikz Should we investigate how to build the `apicast-mapping-service image` or should we let it for the future (as an issue)?

Commit:
re-enabling postCommit in the BuildConfig

Commit:
Adding prometheus exporter endpoint

maneta commented:
@mikz please review 58d79586e9507a6d2fce8ba1e32b1654e0ada872 when you can.

Commit:
updating the postCommit to be more insightfull

Commit:
specifying the port used to scrape
